discovery.sh

```diff
@@ -17,59 +22,51 @@ setup_auto_discovery() {
   log_debug "DEV_NAME=$DEV_NAME"
   log_debug "TOPIC_ROOT=$TOPIC_ROOT"
 
-  echo '{
-   "state_topic": "'${TOPIC_ROOT}'/binary_sensor/presence",
-   "device": {
-    "identifiers": [
-    "'${DEV_ID}'"
-    ],
-    "manufacturer": "tesla-local-control",
-    "model": "Tesla_BLE",
-    "name": "'${DEV_NAME}'",
-    "sw_version": "'${SW_VERSION}'"
-   },
-   "device_class": "presence",
-   "name": "Presence",
-   "unique_id": "'${DEV_ID}'_presence"
-  }' | sed ':a;N;$!ba;s/\n//g' | eval $MOSQUITTO_PUB_BASE -t homeassistant/binary_sensor/${DEV_ID}/presence/config -l
+  log_debug "Leaving setupHAAutoDiscoveryEnvVars() vin:$vin"
 
-  echo '{
-   "command_topic": "'${TOPIC_ROOT}'/config",
-   "device": {
-    "identifiers": [
-    "'${DEV_ID}'"
-    ],
-    "manufacturer": "tesla-local-control",
-    "model": "Tesla_BLE",
-    "name": "'${DEV_NAME}'",
-    "sw_version": "'${SW_VERSION}'"
-   },
-   "device_class": "update",
-   "name": "Generate Keys",
-   "payload_press": "generate-keys",
-   "qos": 1,
-   "unique_id": "'${DEV_ID}'_generate-keys",
-   "entity_category": "config"
-  }' | sed ':a;N;$!ba;s/\n//g' | eval $MOSQUITTO_PUB_BASE -t homeassistant/button/${DEV_ID}/generate-keys/config -l
+}
 
-  echo '{
-   "command_topic": "'${TOPIC_ROOT}'/config",
-   "device": {
-    "identifiers": [
-    "'${DEV_ID}'"
-    ],
-    "manufacturer": "tesla-local-control",
-    "model": "Tesla_BLE",
-    "name": "'${DEV_NAME}'",
-    "sw_version": "'${SW_VERSION}'"
-   },
-   "device_class": "update",
-   "name": "Deploy Key",
-   "payload_press": "deploy-key",
-   "qos": 1,
-   "unique_id": "'${DEV_ID}'_deploy-key",
-   "entity_category": "config"
-  }' | sed ':a;N;$!ba;s/\n//g' | eval $MOSQUITTO_PUB_BASE -t homeassistant/button/${DEV_ID}/deploy-key/config -l
+function setupHAAutoDiscoveryMain() {
+
+  vin=$1
+  keysDir=/share/tesla_ble_mqtt
+
+  log_debug "Entering setupHAAutoDiscoveryMain() vin:$vin"
+
+  setupHAAutoDiscoveryEnvVars $vin
+
+  # If detection is enable, show presence
+  if [ $PRESENCE_DETECTION_TTL -gt 0 ] && [ -n "$BLE_MAC_LIST" ]; then
+    log_debug "setupHAAutoDiscoveryMain() vin:$vin presence detection enable"
+    setupHAPresence $vin
+  fi
+
+  # Newly added car?
+  if [ ! -f $keysDir/${vin}_private.pem ] && [ ! -f $keysDir/${vin}_public.pem ]; then
+
+    log_debug "setupHAAutoDiscoveryMain() vin:$vin newly added car, adding Generate Key menu"
+    # Show button to Generate Keys
+    setupHAGenerateKeys $vin
+
+    # listen_to_mqtt call setupHADeployKey once the keys are generated
+
+  else
+    log_debug "setupHAAutoDiscoveryMain() vin:$vin old car"
+    setupHADeployKey $vin
+    setupHAGenerateKeys $vin
+    setupHAAutoDiscovery $vin
```

Comment:
I suggest only `setupHAAutoDiscovery` there.
No need for the other two if the keys have been generated & deployed

EDIT: maybe keep only DeployKey and remove GenerateKeys

Reply:
I disagree ;) because....

Per the current logic w/ HA Features set enable, each time MQTT gets restarted, HA sends us the topic `.../status = offline`. Once MQTT is back online all the topics vanished. That's why currently we re-run all those. Should we use the retain?  On HA restart the last message would still available but when restarting mqtt, it would be gone.

Only persistent database can help us survive a MQTT restart.

Reply:
Ok understood. Let's do it this way for now